File: apps/php/design-pattern/src/Factory/Factory.php
<?php

declare(strict_types=1);

namespace DesignPattern\Factory;

// --- Animal interface and implementations ---

interface Animal
{
    public function getName(): string;

    public function eat(): string;

    public function speak(): string;
}

class Duck implements Animal
{
    private string $name;

    public function __construct(string $name)
    {
        $this->name = $name;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function eat(): string
    {
        return "{$this->name} はパンを食べています";
    }

    public function speak(): string
    {
        return 'ガーガー';
    }
}

class Frog implements Animal
{
    private string $name;

    public function __construct(string $name)
    {
        $this->name = $name;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function eat(): string
    {
        return "{$this->name} は虫を食べています";
    }

    public function speak(): string
    {
        return 'ケロケロ';
    }
}

class Tiger implements Animal
{
    private string $name;

    public function __construct(string $name)
    {
        $this->name = $name;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function eat(): string
    {
        return "{$this->name} は肉を食べています";
    }

    public function speak(): string
    {
        return 'ガオー';
    }
}

// --- Plant interface and implementations ---

interface Plant
{
    public function getName(): string;

    public function grow(): string;
}

class WaterLily implements Plant
{
    private string $name;

    public function __construct(string $name)
    {
        $this->name = $name;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function grow(): string
    {
        return "{$this->name} は水面に広がっています";
    }
}

class Algae implements Plant
{
    private string $name;

    public function __construct(string $name)
    {
        $this->name = $name;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function grow(): string
    {
        return "{$this->name} は水中で増殖しています";
    }
}

class Tree implements Plant
{
    private string $name;

    public function __construct(string $name)
    {
        $this->name = $name;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function grow(): string
    {
        return "{$this->name} は大きく成長しています";
    }
}

// --- Factory Method pattern: Pond ---

abstract class Pond
{
    private Animal $animal;
    private Plant $plant;

    public function __construct()
    {
        $this->animal = $this->createAnimal();
        $this->plant = $this->createPlant();
    }

    abstract protected function createAnimal(): Animal;

    abstract protected function createPlant(): Plant;

    public function getAnimal(): Animal
    {
        return $this->animal;
    }

    public function getPlant(): Plant
    {
        return $this->plant;
    }

    public function describe(): string
    {
        return sprintf(
            '%s と %s がいます',
            $this->animal->getName(),
            $this->plant->getName()
        );
    }
}

class DuckPond extends Pond
{
    protected function createAnimal(): Animal
    {
        return new Duck('ドナルド');
    }

    protected function createPlant(): Plant
    {
        return new WaterLily('蓮の花');
    }
}

class FrogPond extends Pond
{
    protected function createAnimal(): Animal
    {
        return new Frog('ケロ吉');
    }

    protected function createPlant(): Plant
    {
        return new Algae('藻');
    }
}

// --- Abstract Factory pattern: OrganismFactory + Habitat ---

class OrganismFactory
{
    /** @var class-string<Animal> */
    private string $animalClass;
    /** @var class-string<Plant> */
    private string $plantClass;

    /**
     * @param class-string<Animal> $animalClass
     * @param class-string<Plant> $plantClass
     */
    public function __construct(string $animalClass, string $plantClass)
    {
        $this->animalClass = $animalClass;
        $this->plantClass = $plantClass;
    }

    public function createAnimal(string $name): Animal
    {
        return new ($this->animalClass)($name);
    }

    public function createPlant(string $name): Plant
    {
        return new ($this->plantClass)($name);
    }
}

class Habitat
{
    private Animal $animal;
    private Plant $plant;

    public function __construct(OrganismFactory $factory, string $animalName, string $plantName)
    {
        $this->animal = $factory->createAnimal($animalName);
        $this->plant = $factory->createPlant($plantName);
    }

    public function getAnimal(): Animal
    {
        return $this->animal;
    }

    public function getPlant(): Plant
    {
        return $this->plant;
    }

    public function describe(): string
    {
        return sprintf(
            '%s と %s がいます',
            $this->animal->getName(),
            $this->plant->getName()
        );
    }
}

File: apps/go/design-pattern/factory/factory.go
// Package factory demonstrates the Factory pattern in Go.
// We use interfaces and factory functions instead of abstract classes.
package factory

import "fmt"

// Animal is the product interface for animals.
type Animal interface {
	Name() string
	Eat() string
	Speak() string
}

// Plant is the product interface for plants.
type Plant interface {
	Name() string
	Grow() string
}

// Duck is a concrete animal.
type Duck struct{ DuckName string }

func (d *Duck) Name() string  { return d.DuckName }
func (d *Duck) Eat() string   { return d.DuckName + " は虫を食べる" }
func (d *Duck) Speak() string { return "ガーガー" }

// Frog is a concrete animal.
type Frog struct{ FrogName string }

func (f *Frog) Name() string  { return f.FrogName }
func (f *Frog) Eat() string   { return f.FrogName + " は虫を食べる" }
func (f *Frog) Speak() string { return "ケロケロ" }

// Tiger is a concrete animal.
type Tiger struct{ TigerName string }

func (t *Tiger) Name() string  { return t.TigerName }
func (t *Tiger) Eat() string   { return t.TigerName + " は肉を食べる" }
func (t *Tiger) Speak() string { return "ガオー" }

// WaterLily is a concrete plant.
type WaterLily struct{ LilyName string }

func (w *WaterLily) Name() string { return w.LilyName }
func (w *WaterLily) Grow() string { return w.LilyName + " は水面に広がる" }

// Algae is a concrete plant.
type Algae struct{ AlgaeName string }

func (a *Algae) Name() string { return a.AlgaeName }
func (a *Algae) Grow() string { return a.AlgaeName + " は水中で増殖する" }

// Tree is a concrete plant.
type Tree struct{ TreeName string }

func (t *Tree) Name() string { return t.TreeName }
func (t *Tree) Grow() string { return t.TreeName + " は高く育つ" }

// OrganismFactory creates animals and plants.
type OrganismFactory struct {
	NewAnimal func(name string) Animal
	NewPlant  func(name string) Plant
}

// PondFactory returns a factory for pond organisms.
func PondFactory() *OrganismFactory {
	return &OrganismFactory{
		NewAnimal: func(name string) Animal { return &Duck{DuckName: name} },
		NewPlant:  func(name string) Plant { return &WaterLily{LilyName: name} },
	}
}

// JungleFactory returns a factory for jungle organisms.
func JungleFactory() *OrganismFactory {
	return &OrganismFactory{
		NewAnimal: func(name string) Animal { return &Tiger{TigerName: name} },
		NewPlant:  func(name string) Plant { return &Tree{TreeName: name} },
	}
}

// Habitat uses a factory to create and manage organisms.
type Habitat struct {
	Animals []Animal
	Plants  []Plant
	factory *OrganismFactory
}

// NewHabitat creates a habitat using the given factory.
func NewHabitat(factory *OrganismFactory, animalNames, plantNames []string) *Habitat {
	h := &Habitat{factory: factory}
	for _, name := range animalNames {
		h.Animals = append(h.Animals, factory.NewAnimal(name))
	}
	for _, name := range plantNames {
		h.Plants = append(h.Plants, factory.NewPlant(name))
	}
	return h
}

// Describe returns a description of the habitat.
func (h *Habitat) Describe() string {
	result := ""
	for _, a := range h.Animals {
		result += fmt.Sprintf("%s says %s\n", a.Name(), a.Speak())
	}
	for _, p := range h.Plants {
		result += fmt.Sprintf("%s\n", p.Grow())
	}
	return result
}
